lattices/src/algebra.rs

```diff
@@ -348,6 +348,59 @@ pub fn idempotency<S: Debug + PartialEq + Clone, const N: usize>(
     Ok(())
 }
 
+/// Defines the linearity property
+/// q is linear with respect to some group operation + if q(a+b) = q(a) + q(b)
+/// This is the same as q being a group homomorphism
+// As defined in the paper "DBSP: Automatic Incremental View Maintenance for Rich Query Languages"
+pub fn linearity<S: Debug + PartialEq + Clone, R: Debug + PartialEq + Clone, const N: usize>(
+    items: &[S; N],
+    f: impl Fn(S, S) -> S, // The base operation of the algebraic structure for state
+    g: impl Fn(R, R) -> R, // The base operation of the algebraic structure the query q outputs to
+    q: impl Fn(S) -> R, // The query over f that we want to check for linearity (to incrementalize)
+) -> Result<(), &'static str> {
+    for [a, b] in cartesian_power(items) {
+        if q(f(a.clone(), b.clone())) != g(q(b.clone()), q(a.clone())) {
+            // q(f(a,b)) = f(q(a), q(b))
+            return Err("Linearity check failed.");
+        }
+    }
+    Ok(())
+}
+
+/// Defines the bilinearity property
+/// q is bilinear with respect to + if q(a + b, c) = q(a,c) + q(b,c) and q(a,c + d) = q(a,c) + q(a,d)
+/// This is the same as q being distributive over the addition operation of the three groups S, T, and R in q:S x T --> R
+// As defined in the paper "DBSP: Automatic Incremental View Maintenance for Rich Query Languages"
+pub fn bilinearity<
+    S: Debug + PartialEq + Clone,
+    R: Debug + PartialEq + Clone,
+    T: Debug + PartialEq + Clone,
+    const N_FOR_F: usize,
+    const N_FOR_SECOND_EXAMPLE: usize,
+>(
+    items_f: &[S; N_FOR_F],
+    items_h: &[T; N_FOR_SECOND_EXAMPLE],
```

Comment:
Similarly for here, you can remove the const lengths

Reply:
fixed